File: agent/capabilities/consent_ui.py
"""Endpoint consent and revocation control for remote desktop control.

The agent runs as a Session-0 service and cannot draw an interactive window
itself. This module spawns PowerShell WinForms UI *into the active console
session* using `CreateProcessAsUserW` with a token obtained from
`WTSQueryUserToken(WTSGetActiveConsoleSessionId())` — the same mechanism as
`chat_ui.rs`. Two surfaces live here:

  * a consent dialog (Surface 2) that must be accepted before any input is
    replayed (D-01), naming the requesting admin and tenant (D-03);
  * a persistent stop-control bar (Surface 3) letting the endpoint user end
    control with one click at any time (D-09).

The untrusted requester identity travels to PowerShell only through a
per-session JSON config file read with `ConvertFrom-Json` (T-74-08), never
interpolated into script text or the command line. Nothing is persisted
across sessions (D-04). pywin32 is used for the FFI surface.

The two embedded PowerShell constants live in the sibling `consent_scripts.py`
purely for the 500-line source cap; all Python logic and the public surface
stay here.
"""

import os
import json
import time
import platform
import tempfile
from enum import Enum
from typing import Optional

if platform.system() == "Windows":
    import win32api
    import win32con
    import win32process
    import win32security
    import win32profile
    import win32ts
    import pywintypes
else:
    win32api = win32con = win32process = win32security = win32profile = win32ts = pywintypes = None


class ConsentDecision(Enum):
    Accept = "accept"
    Decline = "decline"
    Timeout = "timeout"


class ConsentError(Exception):
    def __init__(self, reason: str):
        self.reason = reason
        super().__init__(reason)


def _data_dir() -> str:
    """Directory holding per-session config files and decision/stop sentinels."""
    if platform.system() != "Windows":
        return tempfile.gettempdir()
    base = os.environ.get("ProgramData", r"C:\ProgramData")
    return os.path.join(base, "OmniAgent")


def _ensure_scripts_installed() -> None:
    """Write the embedded scripts to data_dir() so spawned processes can launch them. Idempotent."""
    from . import consent_scripts

    d = _data_dir()
    os.makedirs(d, exist_ok=True)
    with open(os.path.join(d, "consent_ui.ps1"), "w") as f:
        f.write(consent_scripts.CONSENT_UI_PS)
    with open(os.path.join(d, "stop_bar.ps1"), "w") as f:
        f.write(consent_scripts.STOP_BAR_PS)


def _parse_decision(body: str) -> Optional[ConsentDecision]:
    """Parse a decision-file body into a ConsentDecision. Unrecognised body returns None."""
    body = body.strip().lower()
    if body == "accept":
        return ConsentDecision.Accept
    if body == "decline":
        return ConsentDecision.Decline
    if body == "timeout":
        return ConsentDecision.Timeout
    return None


def _spawn_in_active_session(cmdline: str) -> None:
    """Launch cmdline as the interactive console user on winsta0\\default (Windows only)."""
    if platform.system() != "Windows":
        raise ConsentError("no interactive desktop session (non-Windows platform)")

    session_id = win32ts.WTSGetActiveConsoleSessionId()
    if session_id == 0xFFFFFFFF:
        raise ConsentError("no active console session (no user logged on)")

    try:
        user_token = win32ts.WTSQueryUserToken(session_id)
    except pywintypes.error as e:
        raise ConsentError(f"WTSQueryUserToken failed ({e.winerror})")

    try:
        primary_token = win32security.DuplicateTokenEx(
            user_token,
            win32con.MAXIMUM_ALLOWED,
            None,
            win32con.SecurityImpersonation,
            win32con.TokenPrimary,
        )
    except pywintypes.error as e:
        win32api.CloseHandle(user_token)
        raise ConsentError(f"DuplicateTokenEx failed ({e.winerror})")
    win32api.CloseHandle(user_token)

    try:
        env = win32profile.CreateEnvironmentBlock(primary_token, False)
        have_env = True
    except pywintypes.error:
        env = None
        have_env = False

    startup = win32process.STARTUPINFO()
    startup.cb = win32process.sizeof(startup)
    startup.lpDesktop = "winsta0\\default"

    flags = win32con.CREATE_UNICODE_ENVIRONMENT | win32con.CREATE_NO_WINDOW

    try:
        hProcess, hThread, _, _ = win32process.CreateProcessAsUser(
            primary_token,
            None,
            cmdline,
            None,
            None,
            False,
            flags,
            env,
            None,
            startup,
        )
    except pywintypes.error as e:
        if have_env and env:
            win32profile.DestroyEnvironmentBlock(env)
        win32api.CloseHandle(primary_token)
        raise ConsentError(f"CreateProcessAsUserW failed ({e.winerror})")

    if have_env and env:
        win32profile.DestroyEnvironmentBlock(env)
    win32api.CloseHandle(primary_token)
    win32api.CloseHandle(hThread)
    win32api.CloseHandle(hProcess)


def request_consent(
    session_id: str,
    requester_name: str,
    requester_email: str,
    tenant_name: str,
    timeout_secs: int = 60,
) -> ConsentDecision:
    """Ask the interactive endpoint user to accept or decline remote control."""
    if platform.system() != "Windows":
        raise ConsentError("no interactive desktop session (non-Windows platform)")

    _ensure_scripts_installed()
    d = _data_dir()

    cfg = {
        "session_id": session_id,
        "requester_name": requester_name,
        "requester_email": requester_email,
        "tenant_name": tenant_name,
        "timeout_secs": timeout_secs,
    }
    cfg_path = os.path.join(d, f"consent_{session_id}.json")
    with open(cfg_path, "w") as f:
        json.dump(cfg, f)

    cmdline = (
        f'powershell.exe -NoProfile -WindowStyle Hidden '
        f'-ExecutionPolicy Bypass -File "{os.path.join(d, "consent_ui.ps1")}" "{cfg_path}"'
    )

    try:
        _spawn_in_active_session(cmdline)
    except ConsentError:
        try:
            os.remove(cfg_path)
        except OSError:
            pass
        raise

    decision_path = os.path.join(d, f"consent_{session_id}.decision")
    deadline = time.time() + timeout_secs
    decision = ConsentDecision.Timeout

    while time.time() < deadline:
        if os.path.exists(decision_path):
            try:
                with open(decision_path, "r") as f:
                    body = f.read()
                parsed = _parse_decision(body)
                if parsed is not None:
                    decision = parsed
                    break
            except OSError:
                pass
        time.sleep(0.2)

    try:
        os.remove(cfg_path)
    except OSError:
        pass
    try:
        os.remove(decision_path)
    except OSError:
        pass

    return decision


def show_stop_bar(session_id: str, requester_name: str) -> None:
    """Show the persistent stop-control bar (Surface 3)."""
    if platform.system() != "Windows":
        raise ConsentError("no interactive desktop session (non-Windows platform)")

    _ensure_scripts_installed()
    d = _data_dir()

    cfg = {"session_id": session_id, "requester_name": requester_name}
    cfg_path = os.path.join(d, f"stop_{session_id}.json")
    with open(cfg_path, "w") as f:
        json.dump(cfg, f)

    cmdline = (
        f'powershell.exe -NoProfile -WindowStyle Hidden '
        f'-ExecutionPolicy Bypass -File "{os.path.join(d, "stop_bar.ps1")}" "{cfg_path}"'
    )

    try:
        _spawn_in_active_session(cmdline)
    except ConsentError:
        try:
            os.remove(cfg_path)
        except OSError:
            pass
        raise


def stop_requested(session_id: str) -> bool:
    """Whether the endpoint user has clicked Stop Control (the .stop sentinel exists)."""
    if platform.system() != "Windows":
        return False
    sentinel = os.path.join(_data_dir(), f"stop_{session_id}.stop")
    return os.path.exists(sentinel)


def hide_stop_bar(session_id: str) -> None:
    """Hide the stop-control bar and clear all session residue."""
    if platform.system() != "Windows":
        return
    d = _data_dir()
    os.makedirs(d, exist_ok=True)

    # Signal the bar to self-close; clear every sentinel/config file.
    close_path = os.path.join(d, f"stop_{session_id}.close")
    try:
        with open(close_path, "w") as f:
            f.write("close")
    except OSError:
        pass

    for suffix in (".stop", ".json", ".close", ".json", ".decision"):
        prefix = "stop_" if not suffix.endswith(".json") or not suffix.endswith(".decision") else "consent_"
        path = os.path.join(d, f"{prefix}{session_id}{suffix}")
        try:
            os.remove(path)
        except OSError:
            pass

File: agent-install/omni-agent-rs/src/consent_ui.rs
//! Session-0 consent dialog and persistent stop-control bar for remote control.
//!
//! The agent runs as a Session-0 service and cannot draw an interactive window
//! itself. This module spawns a PowerShell WinForms consent dialog *into the
//! active console session* using `CreateProcessAsUserW` with a token obtained
//! from `WTSQueryUserToken(WTSGetActiveConsoleSessionId())`. The endpoint user
//! must click Accept before the agent replays any input (D-01), and sees a
//! persistent one-click stop-control bar for as long as control lasts (D-09).
//!
//! The requester identity (name/email/tenant) is untrusted input that reaches
//! this process through a backend payload, so it travels to the PowerShell
//! script ONLY through a per-session JSON config file read with
//! `ConvertFrom-Json` (T-74-08) — never interpolated into the script text or
//! the command line.
//!
//! `winapi` (frozen 0.3 API) is used rather than the `windows` crate to keep
//! the FFI surface stable, mirroring `chat_ui.rs`. The two embedded PowerShell
//! constants live in `consent_scripts.rs` (500-line source-cap split).

use std::collections::HashSet;
use std::sync::{Mutex, OnceLock};

#[cfg(windows)]
use crate::consent_scripts::{CONSENT_UI_PS, STOP_BAR_PS};

/// Session IDs that currently have a consent dialog or stop bar process running.
fn active() -> &'static Mutex<HashSet<String>> {
    static ACTIVE: OnceLock<Mutex<HashSet<String>>> = OnceLock::new();
    ACTIVE.get_or_init(|| Mutex::new(HashSet::new()))
}

fn mark_active(session_id: &str) {
    if let Ok(mut s) = active().lock() {
        s.insert(session_id.to_string());
    }
}

fn mark_closed(session_id: &str) {
    if let Ok(mut s) = active().lock() {
        s.remove(session_id);
    }
}

/// The endpoint user's decision about a control request.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ConsentDecision {
    Accept,
    Decline,
    Timeout,
}

/// Why the consent dialog could not be shown (or the request otherwise failed).
#[derive(Debug)]
pub enum ConsentError {
    /// No interactive console session exists (WTSGetActiveConsoleSessionId
    /// returned 0xFFFF_FFFF) — D-02's refusal condition.
    NoInteractiveSession,
    /// The dialog/bar process could not be spawned.
    SpawnFailed(String),
}

/// Directory that holds the per-session config, decision file, and sentinels.
#[cfg(windows)]
pub fn data_dir() -> std::path::PathBuf {
    let base = std::env::var("ProgramData").unwrap_or_else(|_| r"C:\ProgramData".to_string());
    std::path::Path::new(&base).join("OmniAgent")
}

/// Path where the consent dialog script is materialized.
#[cfg(windows)]
fn script_path() -> std::path::PathBuf {
    data_dir().join("consent_ui.ps1")
}

/// Write the embedded consent-dialog script to ProgramData. Idempotent.
#[cfg(windows)]
fn ensure_script_installed() -> Result<std::path::PathBuf, ConsentError> {
    let dir = data_dir();
    std::fs::create_dir_all(&dir)
        .map_err(|e| ConsentError::SpawnFailed(format!("create {}: {e}", dir.display())))?;
    let p = script_path();
    std::fs::write(&p, CONSENT_UI_PS)
        .map_err(|e| ConsentError::SpawnFailed(format!("write consent script: {e}")))?;
    Ok(p)
}

/// Path where the stop-control bar script is materialized.
#[cfg(windows)]
fn stop_bar_script_path() -> std::path::PathBuf {
    data_dir().join("consent_stop_bar.ps1")
}

/// Write the embedded stop-bar script to ProgramData. Idempotent.
#[cfg(windows)]
fn ensure_stop_bar_installed() -> Result<std::path::PathBuf, ConsentError> {
    let dir = data_dir();
    std::fs::create_dir_all(&dir)
        .map_err(|e| ConsentError::SpawnFailed(format!("create {}: {e}", dir.display())))?;
    let p = stop_bar_script_path();
    std::fs::write(&p, STOP_BAR_PS)
        .map_err(|e| ConsentError::SpawnFailed(format!("write stop bar script: {e}")))?;
    Ok(p)
}

/// Parse the decision file body into a `ConsentDecision`. Only the exact three
/// literal words the script writes parse; anything else (a stale, partial, or
/// tampered file) returns `None` and can never be read as Accept (T-74-17).
pub fn parse_decision(body: &str) -> Option<ConsentDecision> {
    match body.trim() {
        "accept" => Some(ConsentDecision::Accept),
        "decline" => Some(ConsentDecision::Decline),
        "timeout" => Some(ConsentDecision::Timeout),
        _ => None,
    }
}

/// Build the decision-file body for a decision (kept for tests/consistency with
/// the three literal words the script writes).
pub fn decision_payload(decision: &ConsentDecision) -> &'static str {
    match decision {
        ConsentDecision::Accept => "accept",
        ConsentDecision::Decline => "decline",
        ConsentDecision::Timeout => "timeout",
    }
}

/// Serialize the requester identity config JSON for the consent dialog.
#[cfg(windows)]
fn build_consent_config(
    session_id: &str,
    requester_name: &str,
    requester_email: &str,
    tenant_name: &str,
    timeout_secs: u64,
) -> Vec<u8> {
    serde_json::json!({
        "session_id": session_id,
        "requester_name": requester_name,
        "requester_email": requester_email,
        "tenant_name": tenant_name,
        "timeout_secs": timeout_secs,
    })
    .to_string()
    .into_bytes()
}

/// Ask the interactive endpoint user for consent to control their desktop.
///
/// Blocks (synchronously) until the user chooses or the timeout elapses. On the
/// no-interactive-session sentinel it returns `ConsentError::NoInteractiveSession`
/// immediately, before attempting any spawn (D-02). Consumes both the config and
/// the decision file before returning so no residue can be replayed (T-74-17,
/// D-04).
#[cfg(windows)]
pub fn request_consent(
    session_id: &str,
    requester_name: &str,
    requester_email: &str,
    tenant_name: &str,
    timeout_secs: u64,
) -> Result<ConsentDecision, ConsentError> {
    unsafe {
        if win::WTSGetActiveConsoleSessionId() == 0xFFFF_FFFF {
            return Err(ConsentError::NoInteractiveSession);
        }
    }

    let dir = data_dir();
    let script_path = ensure_script_installed()?;

    let cfg = build_consent_config(session_id, requester_name, requester_email, tenant_name, timeout_secs);
    let cfg_path = dir.join(format!("consent_{session_id}.json"));
    std::fs::write(&cfg_path, cfg)
        .map_err(|e| ConsentError::SpawnFailed(format!("write consent config: {e}")))?;

    let cmdline = format!(
        "powershell.exe -NoProfile -WindowStyle Hidden -ExecutionPolicy Bypass -File \"{}\" \"{}\"",
        script_path.display(),
        cfg_path.display()
    );

    if let Err(e) = win::spawn_in_active_session(&cmdline) {
        let _ = std::fs::remove_file(&cfg_path);
        return Err(e);
    }
    mark_active(session_id);

    // Poll the decision file until it appears or the timeout elapses. The script
    // writes one of the three literal words on every exit path.
    let decision_file = dir.join(format!("consent_{session_id}.decision"));
    let deadline = std::time::Instant::now() + std::time::Duration::from_secs(timeout_secs);
    let mut decision = ConsentDecision::Timeout;
    while std::time::Instant::now() < deadline {
        if let Ok(body) = std::fs::read_to_string(&decision_file) {
            if let Some(d) = parse_decision(&body) {
                decision = d;
                break;
            }
            // Unrecognised body — keep polling until the deadline; never Accept.
        }
        std::thread::sleep(std::time::Duration::from_millis(100));
    }

    // Cleanup on every path (T-74-17 / D-04): no residue to replay next session.
    let _ = std::fs::remove_file(&decision_file);
    let _ = std::fs::remove_file(&cfg_path);
    mark_closed(session_id);
    Ok(decision)
}

/// Show the persistent stop-control bar naming the controlling admin.
#[cfg(windows)]
pub fn show_stop_bar(session_id: &str, requester_name: &str) -> Result<(), ConsentError> {
    let dir = data_dir();
    let script_path = ensure_stop_bar_installed()?;

    let cfg = serde_json::json!({
        "session_id": session_id,
        "requester_name": requester_name,
    })
    .to_string()
    .into_bytes();
    let cfg_path = dir.join(format!("consent_{session_id}_bar.json"));
    std::fs::write(&cfg_path, cfg)
        .map_err(|e| ConsentError::SpawnFailed(format!("write stop bar config: {e}")))?;

    let cmdline = format!(
        "powershell.exe -NoProfile -WindowStyle Hidden -ExecutionPolicy Bypass -File \"{}\" \"{}\"",
        script_path.display(),
        cfg_path.display()
    );

    if let Err(e) = win::spawn_in_active_session(&cmdline) {
        let _ = std::fs::remove_file(&cfg_path);
        return Err(e);
    }
    mark_active(session_id);
    Ok(())
}

/// Whether the endpoint user has clicked Stop Control (`stop_{sid}.stop` exists).
pub fn stop_requested(session_id: &str) -> bool {
    #[cfg(windows)]
    {
        data_dir().join(format!("stop_{session_id}.stop")).exists()
    }
    #[cfg(not(windows))]
    {
        let _ = session_id;
        false
    }
}

/// Hide the stop-control bar and clear the session's sentinels/bookkeeping.
#[cfg(windows)]
pub fn hide_stop_bar(session_id: &str) -> Result<(), ConsentError> {
    let dir = data_dir();
    let _ = std::fs::create_dir_all(&dir);
    let close = dir.join(format!("stop_{session_id}.close"));
    std::fs::write(&close, b"close")
        .map_err(|e| ConsentError::SpawnFailed(format!("write stop close sentinel: {e}")))?;
    let _ = std::fs::remove_file(dir.join(format!("stop_{session_id}.stop")));
    let _ = std::fs::remove_file(dir.join(format!("consent_{session_id}_bar.json")));
    mark_closed(session_id);
    Ok(())
}

#[cfg(not(windows))]
pub fn request_consent(
    _session_id: &str,
    _requester_name: &str,
    _requester_email: &str,
    _tenant_name: &str,
    _timeout_secs: u64,
) -> Result<ConsentDecision, ConsentError> {
    Err(ConsentError::NoInteractiveSession)
}

#[cfg(not(windows))]
pub fn show_stop_bar(_session_id: &str, _requester_name: &str) -> Result<(), ConsentError> {
    Err(ConsentError::NoInteractiveSession)
}

#[cfg(not(windows))]
pub fn hide_stop_bar(_session_id: &str) -> Result<(), ConsentError> {
    Ok(())
}

/// Construct the stop sentinel filename for a session (test helper, platform
/// independent — the Windows path uses the same name).
pub fn stop_sentinel_name(session_id: &str) -> String {
    format!("stop_{session_id}.stop")
}

#[cfg(windows)]
mod win {
    use std::os::windows::ffi::OsStrExt;
    use winapi::shared::minwindef::{DWORD, FALSE, LPVOID};
    use winapi::um::errhandlingapi::GetLastError;
    use winapi::um::handleapi::CloseHandle;
    use winapi::um::processthreadsapi::{CreateProcessAsUserW, PROCESS_INFORMATION, STARTUPINFOW};
    use winapi::um::securitybaseapi::DuplicateTokenEx;
    use winapi::um::userenv::{CreateEnvironmentBlock, DestroyEnvironmentBlock};
    use winapi::um::winbase::{WTSGetActiveConsoleSessionId, CREATE_NO_WINDOW, CREATE_UNICODE_ENVIRONMENT};
    use winapi::um::winnt::{
        SecurityImpersonation, TokenPrimary, HANDLE, MAXIMUM_ALLOWED,
    };
    use winapi::um::wtsapi32::WTSQueryUserToken;

    fn wide(s: &str) -> Vec<u16> {
        std::ffi::OsStr::new(s).encode_wide().chain(std::iter::once(0)).collect()
    }

    /// Launch `cmdline` as the interactive console user on winsta0\default.
    pub fn spawn_in_active_session(cmdline: &str) -> Result<(), crate::consent_ui::ConsentError> {
        use crate::consent_ui::ConsentError;
        unsafe {
            let session = WTSGetActiveConsoleSessionId();
            if session == 0xFFFF_FFFF {
                return Err(ConsentError::NoInteractiveSession);
            }

            let mut user_token: HANDLE = std::ptr::null_mut();
            if WTSQueryUserToken(session, &mut user_token) == 0 {
                return Err(ConsentError::SpawnFailed(format!(
                    "WTSQueryUserToken failed ({})", GetLastError()
                )));
            }

            let mut primary: HANDLE = std::ptr::null_mut();
            let dup = DuplicateTokenEx(
                user_token,
                MAXIMUM_ALLOWED,
                std::ptr::null_mut(),
                SecurityImpersonation,
                TokenPrimary,
                &mut primary,
            );
            CloseHandle(user_token);
            if dup == 0 {
                return Err(ConsentError::SpawnFailed(format!(
                    "DuplicateTokenEx failed ({})", GetLastError()
                )));
            }

            let mut env: LPVOID = std::ptr::null_mut();
            let have_env = CreateEnvironmentBlock(&mut env, primary, FALSE) != 0;

            let mut si: STARTUPINFOW = std::mem::zeroed();
            si.cb = std::mem::size_of::<STARTUPINFOW>() as DWORD;
            let mut desktop = wide(r"winsta0\default");
            si.lpDesktop = desktop.as_mut_ptr();

            let mut pi: PROCESS_INFORMATION = std::mem::zeroed();
            let mut cmd = wide(cmdline);
            let flags = CREATE_UNICODE_ENVIRONMENT | CREATE_NO_WINDOW;

            let ok = CreateProcessAsUserW(
                primary,
                std::ptr::null(),
                cmd.as_mut_ptr(),
                std::ptr::null_mut(),
                std::ptr::null_mut(),
                FALSE,
                flags,
                if have_env { env } else { std::ptr::null_mut() },
                std::ptr::null(),
                &mut si,
                &mut pi,
            );
            let err = GetLastError();

            if have_env {
                DestroyEnvironmentBlock(env);
            }
            CloseHandle(primary);

            if ok == 0 {
                return Err(ConsentError::SpawnFailed(format!(
                    "CreateProcessAsUserW failed ({err})"
                )));
            }
            CloseHandle(pi.hThread);
            CloseHandle(pi.hProcess);
            Ok(())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn config_serializes_five_expected_keys() {
        // D-03: the identity config carries all five keys the script reads.
        #[cfg(windows)]
        {
            let cfg = String::from_utf8(build_consent_config("s1", "Ada", "ada@c", "Acme", 60))
                .unwrap();
            let v: serde_json::Value = serde_json::from_str(&cfg).unwrap();
            assert_eq!(v["session_id"], "s1");
            assert_eq!(v["requester_name"], "Ada");
            assert_eq!(v["requester_email"], "ada@c");
            assert_eq!(v["tenant_name"], "Acme");
            assert_eq!(v["timeout_secs"], 60);
        }
        #[cfg(not(windows))]
        {
            assert!(true, "config serialization is Windows-only");
        }
    }

    #[test]
    fn accept_body_parses_to_accept() {
        assert_eq!(parse_decision("accept"), Some(ConsentDecision::Accept));
    }

    #[test]
    fn decline_body_parses_to_decline() {
        assert_eq!(parse_decision("decline"), Some(ConsentDecision::Decline));
    }

    #[test]
    fn timeout_body_parses_to_timeout() {
        assert_eq!(parse_decision("timeout"), Some(ConsentDecision::Timeout));
    }

    #[test]
    fn unknown_decision_body_does_not_parse_to_accept() {
        // T-74-17 / D-04: a stale, partial, or tampered file never reads as Accept.
        for bad in ["", "accept\nignore", "ACCEPT", "maybe", " decline", "{\"accept\":true}", "1"] {
            assert_ne!(parse_decision(bad), Some(ConsentDecision::Accept), "body: {bad:?}");
        }
        // The three valid bodies still round-trip exactly.
        assert_eq!(parse_decision("accept"), Some(ConsentDecision::Accept));
        assert_eq!(parse_decision("decline"), Some(ConsentDecision::Decline));
        assert_eq!(parse_decision("timeout"), Some(ConsentDecision::Timeout));
    }

    #[test]
    fn stop_sentinel_name_construction() {
        // D-09: the stop sentinel the bar writes matches what stop_requested reads.
        assert_eq!(stop_sentinel_name("abc-123"), "stop_abc-123.stop");
        assert_eq!(stop_sentinel_name(""), "stop_.stop");
    }
}
